// @aws-sdk/client-codecommit/dist-es/models/models_1.js
import { CodeCommitServiceException as __BaseException } from "./CodeCommitServiceException";
export class InvalidAuthorArnException extends __BaseException {
    name = "InvalidAuthorArnException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidAuthorArnException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidAuthorArnException.prototype);
    }
}
export class InvalidPullRequestStatusException extends __BaseException {
    name = "InvalidPullRequestStatusException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidPullRequestStatusException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidPullRequestStatusException.prototype);
    }
}
export class InvalidOrderException extends __BaseException {
    name = "InvalidOrderException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidOrderException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidOrderException.prototype);
    }
}
export class InvalidSortByException extends __BaseException {
    name = "InvalidSortByException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidSortByException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidSortByException.prototype);
    }
}
export const OrderEnum = {
    ASCENDING: "ascending",
    DESCENDING: "descending",
};
export const SortByEnum = {
    MODIFIED_DATE: "lastModifiedDate",
    REPOSITORY_NAME: "repositoryName",
};
export class InvalidResourceArnException extends __BaseException {
    name = "InvalidResourceArnException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidResourceArnException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidResourceArnException.prototype);
    }
}
export class ResourceArnRequiredException extends __BaseException {
    name = "ResourceArnRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceArnRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceArnRequiredException.prototype);
    }
}
export class InvalidTargetBranchException extends __BaseException {
    name = "InvalidTargetBranchException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidTargetBranchException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidTargetBranchException.prototype);
    }
}
export class PullRequestApprovalRulesNotSatisfiedException extends __BaseException {
    name = "PullRequestApprovalRulesNotSatisfiedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "PullRequestApprovalRulesNotSatisfiedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PullRequestApprovalRulesNotSatisfiedException.prototype);
    }
}
export class TipOfSourceReferenceIsDifferentException extends __BaseException {
    name = "TipOfSourceReferenceIsDifferentException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TipOfSourceReferenceIsDifferentException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TipOfSourceReferenceIsDifferentException.prototype);
    }
}
export class InvalidOverrideStatusException extends __BaseException {
    name = "InvalidOverrideStatusException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidOverrideStatusException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidOverrideStatusException.prototype);
    }
}
export class OverrideAlreadySetException extends __BaseException {
    name = "OverrideAlreadySetException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "OverrideAlreadySetException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OverrideAlreadySetException.prototype);
    }
}
export class OverrideStatusRequiredException extends __BaseException {
    name = "OverrideStatusRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "OverrideStatusRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OverrideStatusRequiredException.prototype);
    }
}
export class CommentContentRequiredException extends __BaseException {
    name = "CommentContentRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "CommentContentRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, CommentContentRequiredException.prototype);
    }
}
export class CommentContentSizeLimitExceededException extends __BaseException {
    name = "CommentContentSizeLimitExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "CommentContentSizeLimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, CommentContentSizeLimitExceededException.prototype);
    }
}
export class InvalidFileLocationException extends __BaseException {
    name = "InvalidFileLocationException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidFileLocationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidFileLocationException.prototype);
    }
}
export class InvalidFilePositionException extends __BaseException {
    name = "InvalidFilePositionException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidFilePositionException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidFilePositionException.prototype);
    }
}
export class InvalidRelativeFileVersionEnumException extends __BaseException {
    name = "InvalidRelativeFileVersionEnumException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRelativeFileVersionEnumException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRelativeFileVersionEnumException.prototype);
    }
}
export class InvalidReactionValueException extends __BaseException {
    name = "InvalidReactionValueException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidReactionValueException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidReactionValueException.prototype);
    }
}
export class ReactionLimitExceededException extends __BaseException {
    name = "ReactionLimitExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ReactionLimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ReactionLimitExceededException.prototype);
    }
}
export class ReactionValueRequiredException extends __BaseException {
    name = "ReactionValueRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ReactionValueRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ReactionValueRequiredException.prototype);
    }
}
export class FileContentRequiredException extends __BaseException {
    name = "FileContentRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "FileContentRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, FileContentRequiredException.prototype);
    }
}
export class SameFileContentException extends __BaseException {
    name = "SameFileContentException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SameFileContentException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SameFileContentException.prototype);
    }
}
export class InvalidRepositoryTriggerBranchNameException extends __BaseException {
    name = "InvalidRepositoryTriggerBranchNameException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerBranchNameException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerBranchNameException.prototype);
    }
}
export class InvalidRepositoryTriggerCustomDataException extends __BaseException {
    name = "InvalidRepositoryTriggerCustomDataException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerCustomDataException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerCustomDataException.prototype);
    }
}
export class InvalidRepositoryTriggerDestinationArnException extends __BaseException {
    name = "InvalidRepositoryTriggerDestinationArnException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerDestinationArnException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerDestinationArnException.prototype);
    }
}
export class InvalidRepositoryTriggerEventsException extends __BaseException {
    name = "InvalidRepositoryTriggerEventsException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerEventsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerEventsException.prototype);
    }
}
export class InvalidRepositoryTriggerNameException extends __BaseException {
    name = "InvalidRepositoryTriggerNameException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerNameException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerNameException.prototype);
    }
}
export class InvalidRepositoryTriggerRegionException extends __BaseException {
    name = "InvalidRepositoryTriggerRegionException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerRegionException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRepositoryTriggerRegionException.prototype);
    }
}
export class MaximumBranchesExceededException extends __BaseException {
    name = "MaximumBranchesExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "MaximumBranchesExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, MaximumBranchesExceededException.prototype);
    }
}
export class MaximumRepositoryTriggersExceededException extends __BaseException {
    name = "MaximumRepositoryTriggersExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "MaximumRepositoryTriggersExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, MaximumRepositoryTriggersExceededException.prototype);
    }
}
export class RepositoryTriggerBranchNameListRequiredException extends __BaseException {
    name = "RepositoryTriggerBranchNameListRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RepositoryTriggerBranchNameListRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RepositoryTriggerBranchNameListRequiredException.prototype);
    }
}
export class RepositoryTriggerDestinationArnRequiredException extends __BaseException {
    name = "RepositoryTriggerDestinationArnRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RepositoryTriggerDestinationArnRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RepositoryTriggerDestinationArnRequiredException.prototype);
    }
}
export class RepositoryTriggerEventsListRequiredException extends __BaseException {
    name = "RepositoryTriggerEventsListRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RepositoryTriggerEventsListRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RepositoryTriggerEventsListRequiredException.prototype);
    }
}
export class RepositoryTriggerNameRequiredException extends __BaseException {
    name = "RepositoryTriggerNameRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RepositoryTriggerNameRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RepositoryTriggerNameRequiredException.prototype);
    }
}
export class RepositoryTriggersListRequiredException extends __BaseException {
    name = "RepositoryTriggersListRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RepositoryTriggersListRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RepositoryTriggersListRequiredException.prototype);
    }
}
export class TagsMapRequiredException extends __BaseException {
    name = "TagsMapRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TagsMapRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TagsMapRequiredException.prototype);
    }
}
export class InvalidTagKeysListException extends __BaseException {
    name = "InvalidTagKeysListException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidTagKeysListException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidTagKeysListException.prototype);
    }
}
export class TagKeysListRequiredException extends __BaseException {
    name = "TagKeysListRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TagKeysListRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TagKeysListRequiredException.prototype);
    }
}
export class InvalidRuleContentSha256Exception extends __BaseException {
    name = "InvalidRuleContentSha256Exception";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRuleContentSha256Exception",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRuleContentSha256Exception.prototype);
    }
}
export class CommentNotCreatedByCallerException extends __BaseException {
    name = "CommentNotCreatedByCallerException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "CommentNotCreatedByCallerException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, CommentNotCreatedByCallerException.prototype);
    }
}
export class InvalidApprovalStateException extends __BaseException {
    name = "InvalidApprovalStateException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidApprovalStateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidApprovalStateException.prototype);
    }
}
export class MaximumNumberOfApprovalsExceededException extends __BaseException {
    name = "MaximumNumberOfApprovalsExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "MaximumNumberOfApprovalsExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, MaximumNumberOfApprovalsExceededException.prototype);
    }
}
export class PullRequestCannotBeApprovedByAuthorException extends __BaseException {
    name = "PullRequestCannotBeApprovedByAuthorException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "PullRequestCannotBeApprovedByAuthorException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PullRequestCannotBeApprovedByAuthorException.prototype);
    }
}
export class InvalidPullRequestStatusUpdateException extends __BaseException {
    name = "InvalidPullRequestStatusUpdateException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidPullRequestStatusUpdateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidPullRequestStatusUpdateException.prototype);
    }
}
export class PullRequestStatusRequiredException extends __BaseException {
    name = "PullRequestStatusRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "PullRequestStatusRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PullRequestStatusRequiredException.prototype);
    }
}
export class EncryptionKeyRequiredException extends __BaseException {
    name = "EncryptionKeyRequiredException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "EncryptionKeyRequiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, EncryptionKeyRequiredException.prototype);
    }
}
