// renovate/dist/modules/manager/poetry/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.categories = exports.url = exports.supportsLockFileMaintenance = exports.updateLockedDependency = exports.updateArtifacts = exports.extractPackageFile = exports.bumpPackageVersion = void 0;
const git_refs_1 = require("../../datasource/git-refs");
const git_tags_1 = require("../../datasource/git-tags");
const github_releases_1 = require("../../datasource/github-releases");
const github_tags_1 = require("../../datasource/github-tags");
const gitlab_tags_1 = require("../../datasource/gitlab-tags");
const pypi_1 = require("../../datasource/pypi");
var update_1 = require("../pep621/update");
Object.defineProperty(exports, "bumpPackageVersion", { enumerable: true, get: function () { return update_1.bumpPackageVersion; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var update_locked_1 = require("./update-locked");
Object.defineProperty(exports, "updateLockedDependency", { enumerable: true, get: function () { return update_locked_1.updateLockedDependency; } });
exports.supportsLockFileMaintenance = true;
exports.url = 'https://python-poetry.org/docs';
exports.categories = ['python'];
exports.defaultConfig = {
    fileMatch: ['(^|/)pyproject\\.toml$'],
};
exports.supportedDatasources = [
    pypi_1.PypiDatasource.id,
    github_tags_1.GithubTagsDatasource.id,
    github_releases_1.GithubReleasesDatasource.id,
    gitlab_tags_1.GitlabTagsDatasource.id,
    git_refs_1.GitRefsDatasource.id,
    git_tags_1.GitTagsDatasource.id,
];
