// renovate/dist/modules/manager/copier/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.url = exports.updateDependency = exports.extractPackageFile = exports.updateArtifacts = void 0;
const tslib_1 = require("tslib");
const git_tags_1 = require("../../datasource/git-tags");
const pep440 = tslib_1.__importStar(require("../../versioning/pep440"));
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var update_1 = require("./update");
Object.defineProperty(exports, "updateDependency", { enumerable: true, get: function () { return update_1.updateDependency; } });
exports.url = 'https://copier.readthedocs.io';
exports.defaultConfig = {
    fileMatch: ['(^|/)\\.copier-answers(\\..+)?\\.ya?ml'],
    versioning: pep440.id,
};
exports.supportedDatasources = [git_tags_1.GitTagsDatasource.id];
