// renovate/dist/modules/manager/pep621/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.categories = exports.url = exports.displayName = exports.supportsLockFileMaintenance = exports.updateArtifacts = exports.extractPackageFile = exports.bumpPackageVersion = void 0;
const pypi_1 = require("../../datasource/pypi");
var update_1 = require("./update");
Object.defineProperty(exports, "bumpPackageVersion", { enumerable: true, get: function () { return update_1.bumpPackageVersion; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
exports.supportsLockFileMaintenance = true;
exports.displayName = 'PEP 621';
exports.url = 'https://peps.python.org/pep-0621';
exports.categories = ['python'];
exports.defaultConfig = {
    fileMatch: ['(^|/)pyproject\\.toml$'],
};
exports.supportedDatasources = [pypi_1.PypiDatasource.id];
