// renovate/dist/modules/manager/helmfile/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.categories = exports.url = exports.supportsLockFileMaintenance = exports.updateArtifacts = exports.extractPackageFile = void 0;
const docker_1 = require("../../datasource/docker");
const helm_1 = require("../../datasource/helm");
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
exports.supportsLockFileMaintenance = true;
exports.url = 'https://helmfile.readthedocs.io';
exports.categories = ['cd', 'helm', 'kubernetes'];
exports.defaultConfig = {
    registryAliases: {
        stable: 'https://charts.helm.sh/stable',
    },
    commitMessageTopic: 'helm chart {{depName}}',
    fileMatch: ['(^|/)helmfile\\.ya?ml(?:\\.gotmpl)?$'],
};
exports.supportedDatasources = [helm_1.HelmDatasource.id, docker_1.DockerDatasource.id];
