// renovate/dist/modules/manager/conan/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.categories = exports.url = exports.supportsLockFileMaintenance = exports.getRangeStrategy = exports.updateArtifacts = exports.extractPackageFile = void 0;
const tslib_1 = require("tslib");
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var range_1 = require("./range");
Object.defineProperty(exports, "getRangeStrategy", { enumerable: true, get: function () { return range_1.getRangeStrategy; } });
const conan_1 = require("../../datasource/conan");
const conan = tslib_1.__importStar(require("../../versioning/conan"));
exports.supportsLockFileMaintenance = true;
exports.url = 'https://docs.conan.io';
exports.categories = ['c'];
exports.defaultConfig = {
    fileMatch: ['(^|/)conanfile\\.(txt|py)$'],
    datasource: conan_1.ConanDatasource.id,
    versioning: conan.id,
    enabled: false, // See https://github.com/renovatebot/renovate/issues/14170
};
exports.supportedDatasources = [conan_1.ConanDatasource.id];
