// renovate/dist/modules/manager/bundler/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.categories = exports.url = exports.updateLockedDependency = exports.updateArtifacts = exports.extractPackageFile = exports.supportsLockFileMaintenance = void 0;
const tslib_1 = require("tslib");
const ruby_version_1 = require("../../datasource/ruby-version");
const rubygems_1 = require("../../datasource/rubygems");
const rubyVersioning = tslib_1.__importStar(require("../../versioning/ruby"));
const artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
const extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
const update_locked_1 = require("./update-locked");
Object.defineProperty(exports, "updateLockedDependency", { enumerable: true, get: function () { return update_locked_1.updateLockedDependency; } });
exports.supportsLockFileMaintenance = true;
exports.url = 'https://bundler.io/docs.html';
exports.categories = ['ruby'];
exports.defaultConfig = {
    fileMatch: ['(^|/)Gemfile$'],
    versioning: rubyVersioning.id,
};
exports.supportedDatasources = [
    rubygems_1.RubygemsDatasource.id,
    ruby_version_1.RubyVersionDatasource.id,
];
