// renovate/dist/modules/manager/gleam/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.supportsLockFileMaintenance = exports.defaultConfig = exports.url = exports.getRangeStrategy = exports.updateArtifacts = exports.extractPackageFile = void 0;
const tslib_1 = require("tslib");
const hex_1 = require("../../datasource/hex");
const hexVersioning = tslib_1.__importStar(require("../../versioning/hex"));
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var range_1 = require("./range");
Object.defineProperty(exports, "getRangeStrategy", { enumerable: true, get: function () { return range_1.getRangeStrategy; } });
exports.url = 'https://gleam.run/documentation';
exports.defaultConfig = {
    fileMatch: ['(^|/)gleam.toml$'],
    versioning: hexVersioning.id,
};
exports.supportsLockFileMaintenance = true;
exports.supportedDatasources = [hex_1.HexDatasource.id];
