// renovate/dist/constants/platforms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BITBUCKET_SERVER_API_USING_HOST_TYPES = exports.BITBUCKET_API_USING_HOST_TYPES = exports.GITLAB_API_USING_HOST_TYPES = exports.GITHUB_API_USING_HOST_TYPES = exports.GITEA_API_USING_HOST_TYPES = exports.PLATFORM_HOST_TYPES = void 0;
exports.PLATFORM_HOST_TYPES = [
    'azure',
    'bitbucket',
    'bitbucket-server',
    'codecommit',
    'gerrit',
    'gitea',
    'github',
    'gitlab',
    'local',
];
exports.GITEA_API_USING_HOST_TYPES = [
    'gitea',
    'gitea-changelog',
    'gitea-releases',
    'gitea-tags',
];
exports.GITHUB_API_USING_HOST_TYPES = [
    'github',
    'github-releases',
    'github-release-attachments',
    'github-tags',
    'pod',
    'hermit',
    'github-changelog',
    'conan',
];
exports.GITLAB_API_USING_HOST_TYPES = [
    'gitlab',
    'gitlab-releases',
    'gitlab-tags',
    'gitlab-packages',
    'gitlab-changelog',
    'pypi',
];
exports.BITBUCKET_API_USING_HOST_TYPES = [
    'bitbucket',
    'bitbucket-changelog',
    'bitbucket-tags',
];
exports.BITBUCKET_SERVER_API_USING_HOST_TYPES = [
    'bitbucket-server',
    'bitbucket-server-changelog',
    'bitbucket-server-tags',
];
