// renovate/dist/modules/platform/local/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.id = void 0;
exports.initPlatform = initPlatform;
exports.getRepos = getRepos;
exports.initRepo = initRepo;
exports.findIssue = findIssue;
exports.getIssueList = getIssueList;
exports.getRawFile = getRawFile;
exports.getJsonFile = getJsonFile;
exports.getPrList = getPrList;
exports.ensureIssueClosing = ensureIssueClosing;
exports.ensureIssue = ensureIssue;
exports.massageMarkdown = massageMarkdown;
exports.maxBodyLength = maxBodyLength;
exports.updatePr = updatePr;
exports.mergePr = mergePr;
exports.addReviewers = addReviewers;
exports.addAssignees = addAssignees;
exports.createPr = createPr;
exports.deleteLabel = deleteLabel;
exports.setBranchStatus = setBranchStatus;
exports.getBranchStatus = getBranchStatus;
exports.getBranchStatusCheck = getBranchStatusCheck;
exports.ensureCommentRemoval = ensureCommentRemoval;
exports.ensureComment = ensureComment;
exports.getPr = getPr;
exports.findPr = findPr;
exports.getBranchPr = getBranchPr;
exports.id = 'local';
function initPlatform(params) {
    return Promise.resolve({
        dryRun: 'lookup',
        endpoint: 'local',
        persistRepoData: true,
        requireConfig: 'optional',
    });
}
function getRepos() {
    return Promise.resolve([]);
}
function initRepo() {
    return Promise.resolve({
        defaultBranch: '',
        isFork: false,
        repoFingerprint: '',
    });
}
function findIssue() {
    return Promise.resolve(null);
}
function getIssueList() {
    return Promise.resolve([]);
}
function getRawFile() {
    return Promise.resolve(null);
}
function getJsonFile() {
    return Promise.resolve(null);
}
function getPrList() {
    return Promise.resolve([]);
}
function ensureIssueClosing() {
    return Promise.resolve();
}
function ensureIssue() {
    return Promise.resolve(null);
}
function massageMarkdown(input) {
    return input;
}
/**
 * Unsed, no Dashboard
 */
function maxBodyLength() {
    return Infinity;
}
function updatePr() {
    return Promise.resolve();
}
function mergePr() {
    return Promise.resolve(false);
}
function addReviewers() {
    return Promise.resolve();
}
function addAssignees() {
    return Promise.resolve();
}
function createPr() {
    return Promise.resolve(null);
}
function deleteLabel() {
    return Promise.resolve();
}
function setBranchStatus() {
    return Promise.resolve();
}
function getBranchStatus() {
    return Promise.resolve('red');
}
function getBranchStatusCheck() {
    return Promise.resolve(null);
}
function ensureCommentRemoval() {
    return Promise.resolve();
}
function ensureComment() {
    return Promise.resolve(false);
}
function getPr() {
    return Promise.resolve(null);
}
function findPr() {
    return Promise.resolve(null);
}
function getBranchPr() {
    return Promise.resolve(null);
}
