// renovate/dist/modules/manager/hermit/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.defaultConfig = exports.url = exports.updateDependency = exports.extractPackageFile = exports.updateArtifacts = void 0;
const hermit_1 = require("../../datasource/hermit");
const hermit_2 = require("../../versioning/hermit");
const default_config_1 = require("./default-config");
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var update_1 = require("./update");
Object.defineProperty(exports, "updateDependency", { enumerable: true, get: function () { return update_1.updateDependency; } });
exports.url = 'https://cashapp.github.io/hermit';
exports.defaultConfig = {
    fileMatch: default_config_1.defaultConfig.fileMatch,
    excludeCommitPaths: default_config_1.defaultConfig.excludeCommitPaths,
    versioning: hermit_2.id,
};
exports.supportedDatasources = [hermit_1.HermitDatasource.id];
