// renovate/dist/modules/datasource/aws-eks-addon/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksAddonsFilter = exports.EksAddonsFilterSchema = void 0;
const zod_1 = require("zod");
const regex_1 = require("../../../util/regex");
const schema_utils_1 = require("../../../util/schema-utils");
const stringIsBoolSchema = zod_1.z
    .string()
    .transform((value) => {
    if (value === 'true') {
        return true;
    }
    else if (value === 'false') {
        return false;
    }
    else {
        return false;
    }
});
exports.EksAddonsFilterSchema = zod_1.z.object({
    addonName: zod_1.z.string().nonempty().regex((0, regex_1.regEx)('^[a-z0-9][a-z0-9-]*[a-z0-9]$')),
    kubernetesVersion: zod_1.z
        .string()
        .regex((0, regex_1.regEx)('^(?<major>\\d+)\\.(?<minor>\\d+)$'))
        .optional(),
    default: zod_1.z.oboolean().or(stringIsBoolSchema),
    region: zod_1.z.string().optional(),
    profile: zod_1.z.string().optional(),
});
exports.EksAddonsFilter = schema_utils_1.Json.pipe(exports.EksAddonsFilterSchema);
