// good-enough-parser/dist/cjs/query/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRoot = exports.str = exports.tree = exports.alt = exports.opt = exports.many = exports.num = exports.comment = exports.op = exports.sym = exports.join = exports.handler = exports.begin = void 0;
var builder_1 = require("./builder");
Object.defineProperty(exports, "begin", { enumerable: true, get: function () { return builder_1.begin; } });
Object.defineProperty(exports, "handler", { enumerable: true, get: function () { return builder_1.handler; } });
Object.defineProperty(exports, "join", { enumerable: true, get: function () { return builder_1.join; } });
Object.defineProperty(exports, "sym", { enumerable: true, get: function () { return builder_1.sym; } });
Object.defineProperty(exports, "op", { enumerable: true, get: function () { return builder_1.op; } });
Object.defineProperty(exports, "comment", { enumerable: true, get: function () { return builder_1.comment; } });
Object.defineProperty(exports, "num", { enumerable: true, get: function () { return builder_1.num; } });
Object.defineProperty(exports, "many", { enumerable: true, get: function () { return builder_1.many; } });
Object.defineProperty(exports, "opt", { enumerable: true, get: function () { return builder_1.opt; } });
Object.defineProperty(exports, "alt", { enumerable: true, get: function () { return builder_1.alt; } });
Object.defineProperty(exports, "tree", { enumerable: true, get: function () { return builder_1.tree; } });
Object.defineProperty(exports, "str", { enumerable: true, get: function () { return builder_1.str; } });
Object.defineProperty(exports, "buildRoot", { enumerable: true, get: function () { return builder_1.buildRoot; } });
__exportStar(require("./types"), exports);
