// renovate/dist/modules/manager/mix/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.supportsLockFileMaintenance = exports.defaultConfig = exports.categories = exports.url = exports.getRangeStrategy = exports.updateArtifacts = exports.extractPackageFile = void 0;
const git_tags_1 = require("../../datasource/git-tags");
const github_tags_1 = require("../../datasource/github-tags");
const hex_1 = require("../../datasource/hex");
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var range_1 = require("./range");
Object.defineProperty(exports, "getRangeStrategy", { enumerable: true, get: function () { return range_1.getRangeStrategy; } });
exports.url = 'https://hexdocs.pm/mix/Mix.html';
exports.categories = ['elixir'];
exports.defaultConfig = {
    fileMatch: ['(^|/)mix\\.exs$'],
};
exports.supportsLockFileMaintenance = true;
exports.supportedDatasources = [
    github_tags_1.GithubTagsDatasource.id,
    git_tags_1.GitTagsDatasource.id,
    hex_1.HexDatasource.id,
];
